### DISCUSSION TITLE
Are Mongoid transactions actually thread-safe?

### DISCUSSION THREAD

**Discussion by arg** [mongodb/mongoid#5870] [Q&A] [answered]
Our Rails app runs in a pretty common multi-threaded environment powered by Puma. After the recent upgrade to Mongoid v9, we'd like to start using this [new block-syntax](https://www.mongodb.com/docs/mongoid/current/reference/transactions/#higher-level-api) for transactions. However after reading the documentation, I've got a concern about feature's thread safety. Here are few quotes:

> When the transaction method is called, Mongoid does the following:
> * **creates a session** on a client that is used by the receiver of the transaction method call;
>
> ([source](https://www.mongodb.com/docs/mongoid/current/reference/transactions/#higher-level-api))

and

> Please note the following limitations of sessions:
> * Sessions cannot be shared across threads; **sessions are not thread-safe**.
>
> ([source](https://www.mongodb.com/docs/mongoid/current/reference/sessions/#sessions)) 

Does it mean that Mongoid transactions are not thread-safe and therefore should not be used in a multi-threaded setups, such as Puma?

Thanks

**Comment by comandeo-mongo** [comment, marked as answer]
Hi @arg , thank you for your question! Mongoid transactions API is thread safe. When you call `#with_session` method, it creates a new session for the current persistence context, and stores it in the `Thread.local` storage ( https://github.com/mongodb/mongoid/blob/master/lib/mongoid/clients/sessions.rb#L47). This session will be used inside the block. 

The new `#transaction` method uses `#with_session` under the hood, so the same applies here.

Therefore, you are safe unless you explicitly pass the session to another thread. This won't work, and should be avoided.

I would like to point out that `#with_session` method is not the new transaction API introduced in 9. If you were using transactions in 8.x, and didn't experience any issues, it will be the same in 9.x. 

I hope this clarifies the situation for you. Please reach us out if you still have question, or if you experience issues with the new transactions API.

Thank you for using MongoDB and Mongoid!

**Reply by arg** [reply]
@comandeo-mongo thank you for clarifying.
